### ISSUE TITLE
Api discuss

### ISSUE THREAD

**Issue by djx314** [zio/zio#4187]
https://github.com/zio/zio/blob/f8f72fc73f4edb9a5711769b3299b8e536916a18/core/shared/src/main/scala/zio/ZIO.scala#L318

I am writing a zio api lib use nat programming. In ZIO's catchAll, I find that some api is not so friendly. In my code from djx314/XIO

```scala
val live: XLayer[xlog.XLogging#Plus[XEnv], XError1[Throwable], DB] = XLayer.fromManagedMany(
    XManaged.make(
      for {
        _  <- xlog.info("db start")
        db <- xblocking.effectBlocking(database1)
      } yield XHas1(DB1Content(db))
    )(db =>
      for {
        u <- xblocking
          .effectBlocking(db._0.get.db.close())
          .retryN(3)
          .catchAll(e => xlog.error("db close error", Cause.die(e.sureRight)): XIO[xlog.XLogging#Plus[XEnv], XError0, Unit])
        _ <- xlog.info("db close")
      } yield u: Any))
```

```scala
XLayer[xlog.XLogging#Plus[XEnv], XError1[Throwable], DB]
```
means that
```scala
ZLayer[Logging with Env, Throwable, DB]
```
`catchAll` means that
```scala
blocking.effectBlocking(db._0.get.db.close())
  .retryN(3)
  .catchAll(e => log.error(db close error, Cause.die(e)): ZIO[Logging with Env, Nothing, Unit])
```
I means that why not change `catchAll` to
```scala
final def catchAll[R1, E2, A1 >: A](h: E => ZIO[R1, E2, A1])(implicit ev: CanFail[E]): ZIO[R1 with R, E2, A1]
```
that avoid the `xxx: ZIO[Logging with Env, Nothing, Unit]`
Just
```scala
blocking.effectBlocking(db._0.get.db.close())
  .retryN(3)
  .catchAll(e => log.error(db close error, Cause.die(e))
```
is OK.

**Comment by adamgfraser** [comment]
@djx314 Hi! Would it be possible for you to minimize your example a bit, ideally to something that only uses ZIO directly? I tried this and it worked correctly:

```scala
import zio._
import zio.blocking.Blocking

object Example {
  
  trait Database
  
  trait Env
  trait Logging
  
  def logError[E](e: E): ZIO[Logging, Nothing, Unit] =
    ???
  
  val effect: ZIO[Env with Logging with Blocking, Throwable, Unit] =
    blocking.effectBlocking(new Database {})
      .retryN(3)
      .catchAll(e => logError(e))
      .unit
}
```

In general the `R with `R1` pattern can be used instead of the `R1 <: R` pattern and we have used it in some cases, but we have found that the second pattern can lead to simpler inferred types, in particular avoiding `R with Logging with Logging with Logging` in inferred types. I haven't seen type inferences issues with `R1 <: R` since with `R` being contravariant the type can always be narrowed to be a subtype of `R`, getting to the intersection type. But clearly you are having issues so want to make sure I can reproduce so we can understand and address.

Thank you again for raising!
